elliottlib/cli/find_builds_cli.py

```diff
@@ -307,49 +289,129 @@ def _find_shipped_builds(build_ids: List[Union[str, int]], brew_session: koji.Cl
+    unknown_tags_builds = [b for b in builds if "_tags" not in b]  # finds builds whose tags are not cached
+    build_tag_lists = brew.get_builds_tags(unknown_tags_builds, brew_session)
+    for build, tags in zip(unknown_tags_builds, build_tag_lists):
+        build["_tags"] = {tag['name'] for tag in tags}
+    # Finds and sets the accepted tag (rhaos-x.y-rhel-z-[candidate|hotfix]) for each build
+    for build in builds:
+        accepted_tag = next(filter(lambda tag: tag in tag_pv_map, build["_tags"]), None)
+        if not accepted_tag:
+            raise IOError(f"Build {build['nvr']} has Brew tags {tags}, but none of them has an Errata product version mapping.")
+        build["tag_name"] = accepted_tag
+
+
+def _fetch_builds_by_kind_rpm(runtime: Runtime, assembly: Optional[str], tag_pv_map: Dict[str, str], brew_event: Optional[int], brew_session: koji.ClientSession):
     green_prefix('Generating list of rpms: ')
     click.echo('Hold on a moment, fetching Brew builds')
-    tag_component_tuples = [(tag, None) for tag in tag_pv_map]
-    brew_builds = brew.get_tagged_builds(tag_component_tuples, "rpm", event=brew_event, session=brew_session)
-    brew_latest_builds = list(_find_latest_builds(brew_builds, runtime.assembly))
 
-    shipped = _find_shipped_builds([b["id"] for b in brew_latest_builds], brew_session)
-    unshipped = [b for b in brew_latest_builds if b["id"] not in shipped]
+    qualified_builds: Set[int, Dict] = {}  # keys are build ids, values are brew build dicts
+    builder = BuildFinder(brew_session, logger=LOGGER)
+    attachable_components = set()  # a set of RPM component names; "attachable" means the component is whitelisted to attach to OCP advisories.
+    not_attachable_nvrs: Set[str] = set()
+
+    def add_builds(component_builds):
+        for component, build in component_builds.items():
+            if component not in attachable_components:
+                not_attachable_nvrs.add(build["nvr"])
+            elif build["id"] not in qualified_builds:
+                qualified_builds[build["id"]] = build
+
+    for tag in tag_pv_map:
+        # keys are rpm component names, values are nvres
+        component_builds: Dict[str, Dict] = builder.from_tag("rpm", tag, inherit=True, assembly=assembly, event=brew_event)
+        attachable_components |= component_builds.keys()  # components from our tag are always attachable
+
+        if runtime.assembly_basis_event:
+            # If an assembly has a basis event, rpms pinned by "is" and group dependencies should take precedence over every build from the tag
+            el_version = isolate_el_version_in_brew_tag(tag)
+            if not el_version:
+                continue   # Only honor pinned rpms if this tag is relevant to a RHEL version
+
+            # Honors pinned NVRs by "is"
+            pinned_by_is = builder.from_pinned_by_is(el_version, runtime.assembly, runtime.get_releases_config(), runtime.rpm_map)
```

Comment:
Yeah it is kinda duplicate because it is also taken from plashet and we were developing in parallel.
But I would prefer to use `from_pinned_by_is` for querying rpms because:
1. It can use the cache built by `listTagged` from the previous step;
2. Avoid Brew API calls for rpm configs that don't have an `is` entry;
3. The return value of `from_pinned_by_is` is of type `Dict[component_name, build_dict]`, which is easier to take precedence over any NVR for the same component found from the tag.

Reply:
Replicating logic is nearly always more expensive in the long run. If we change how `is` is interpreted at some point in the future, having only one method which houses the logic makes the change trivial.
That's said, we can work on unifying in a separate story once we have the MVP.

Reply:
I will make a JIRA story for unifying Replicating logic like this and those shared in elliott and doozer.

Reply:
I added some comments to https://issues.redhat.com/browse/ART-3106.